feat: svg sprite generation

amtins opened the pull request:
## Description

Adds SVG sprite generation to customize video.js icons.

## Specific Changes proposed

- The `bin` folder contains the `cli.js` file for command-line usage.
- The `icons` folder contains all the SVG icons used to generate the sprite.
- The `src` folder contains the `svg-sprite` and `SVGO` configuration files, as well as the script for generating the sprite.
- The `template` folder includes the HTML file used for the preview page, along with a configuration file that can be generated via the CLI.
- The `vjs-svg-sprite` folder contains the preview page and the default sprite.
- The `index.js` file is only used for development and generating the default sprite.
- The `svg-icons.json` file defines the default configuration.
- The .nvmrc file to specify node version

amtins committed:
feat: svg sprite generation

Adds SVG sprite generation to customize video.js icons

- The `bin` folder contains the `cli.js` file for command-line usage.
- The `icons` folder contains all the SVG icons used to generate the sprite.
- The `src` folder contains the `svg-sprite` and `SVGO` configuration files, as well as the script for generating the sprite.
- The `template` folder includes the HTML file used for the preview page, along with a configuration file that can be generated via the CLI.
- The `vjs-svg-sprite` folder contains the preview page and the default sprite.
- The `index.js` file is only used for development and generating the default sprite.
- The `svg-icons.json` file defines the default configuration.

mister-ben commented:
Great stuff, a couple of quick thoughts

* A `.nvmrc` would be good
* I notice the output is a little smaller than what we currently have in video.js/src/images/icons.svg. Is this just better optimised?
* It's not too important as long as it is documented, but should the output dir be named something more conventional like `dist`?

amtins committed:
chore: add .nvmrc file to specify node version

amtins commented:
@mister-ben I've added the .nvmrc file.

> I notice the output is a little smaller than what we currently have in video.js/src/images/icons.svg. Is this just better optimised?

Yes, because I've applied two optimizations, the first directly to the icons and the second directly to the generated sprite, which should further reduce the size of the final sprite.

**Individual icon optimization**
https://github.com/videojs/svg-sprite-generator/blob/f9b1014848608383794a8b9aa6a61b535711cfcf/src/vjs-sprite.js#L41

**Sprite optimization**
https://github.com/videojs/svg-sprite-generator/blob/f9b1014848608383794a8b9aa6a61b535711cfcf/src/vjs-sprite.js#L53

> It's not too important as long as it is documented, but should the output dir be named something more conventional like `dist`?

Yes, I can also change the destination filename to `dist`. While we're on the subject, do you have any suggestions for the final filename? To import the sprite, you have to do `@videojs/svg-sprite-generator/dist/vjs-sprite-icons.svg`. In retrospect, I have mixed feelings about the filename; I think there are still too many characters.

amtins committed:
fix(template): footer link pointing to wrong repo

Regenerate `vjs-svg-sprite/index.html` to remove reference to wrong repo.

amtins committed:
chore(svg-icons): rename output directory

From now on, the value of `output-dir` is `dist` instead of `vjs-svg-sprite`.

Access to the SVG file will be via the path
`@videojs/svg-sprite-generator/dist/vjs-sprite-icons.svg` instead of
`@videojs/svg-sprite-generator/vjs-svg-sprite/vjs-sprite-icons.svg`.

wseymour15 commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2024 amtins<amtins.dev@gmail.com>.
```
We may want to update this file to align with https://github.com/videojs/video.js/blob/main/LICENSE

wseymour15 commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+vjs-sprite-tmp_*
```
Do we want to include `dist` in this file?

amtins replied:
Ideally, yes, in order to replace the [player.js](https://github.com/videojs/video.js/blob/19ca3f2ebe0ec7f0374fa3bc7df1457f9f185823/src/js/player.js#L36) import with the file generated by this project. A bit like [videojs-icons.css](https://github.com/videojs/font/blob/master/css/videojs-icons.css)

wseymour15 commented on src/svgo.clean.config.js:
```diff
@@ -0,0 +1,27 @@
+      params: {
+        attributes: [
+          {
+            style: 'display: none'
```
I am sure there is good reason, but why do we want to set this style on clean?

amtins replied:
The main reason is that svg-sprite adds `style="position:absolute"`. So the cleanup is to replace `position:absolute` with `display:none`. Currently, `display:none` is applied via JavaScript and is not present in the `icons.svg` file. This would allow to remove [line 545](https://github.com/videojs/video.js/blob/19ca3f2ebe0ec7f0374fa3bc7df1457f9f185823/src/js/player.js#L545).

wseymour15 commented on template/svg.config.file.js:
```diff
@@ -0,0 +1,52 @@
+  "root-dir": "",
+  "output-dir": "",
+  "icons": {
+    "audio-description": "",
```
This seems bit repetitive as a similar structure exists in scg-icons.json. Would it be possible to just move those values here and remove the other file?

It may not be possible as I am unfamiliar with `svg-sprite`, but figured I would mention it and try to get a better understanding.

amtins replied:
The lack of a `README` adds some confusion. To give a bit of context, there are two aspects to this project:

1. Standardize and simplify the generation of the SVG sprite used by video.js. This would allow us to remove the [`icons.svg`](https://github.com/videojs/video.js/blob/main/src/images/icons.svg?short_path=31700ff) file and replace the import with the file generated by this project. I was inspired by the same philosophy as the [font project](https://github.com/videojs/font), which is why this project also provides the final file. So the `svg-icons.json` file is used to generate the sprite for this project.
2. Allow developers to easily replace the default icons used by video.js. All of this can be done using the command-line tool. Therefore, `svg.config.file.js` will serve as a configuration file template for developers to customize the icons.

wseymour15 commented on package.json:
```diff
@@ -0,0 +1,34 @@
+    "vjs-svg-sprite": "./bin/cli.js"
+  },
+  "scripts": {
+    "build": "node index.js",
```
I am trying to think if there are any other scripts we may want to include here.. Would we want to include `lint`, or maybe `clean` to remove the dist?

amtins replied:
I have added the `linter` as well as `husky`. 

Thank you for taking the time to review this PR and for the comments. ❤️ 

I am open to any other suggestions or ideas you may have.

wseymour15 reviewed (commented):
This looks great, nice work! I left a few comments and questions.

Regarding the usage of this repo, is the distributed svg file meant to replace the file here? https://github.com/videojs/video.js/pull/8260/files#diff-56a92727a19631fea08e4d8424bbf11cbf939176723287f3e2947eded506fc83

amtins committed:
chore(license): content update

amtins committed:
chore(package): update license type

amtins committed:
chore(npm): add videojs-standard

amtins committed:
chore(npm): add husky

Run linter on pre-commit hook

amtins committed:
style(cli): comply with linter

amtins committed:
style(index): comply with linter

amtins committed:
style(svgo-clean-config): comply with linter

amtins committed:
style(svgo-config): comply with linter

amtins committed:
style(svg-config-file): comply with linter

amtins committed:
style(vjs-sprite): comply with linter

amtins committed:
style(svg-config-file): comply with linter

amtins committed:
chore(npm): lint errors

amtins committed:
chore(npm): vjsstandard ignore file

The svg-sprite.config.js file is ignored because import.meta.url is not
compatible with the version of javascript used by vjsstandard.

amtins committed:
style(svg-sprite-config): use quotes instead double quotes

amtins committed:
chore(npm): fix start script path

amtins committed:
fix(icons): use the X icon for twitter

Regenerate sprite file and html page.
